plugins/lua/buildingplan/unlink_mechanisms.lua

```diff
@@ -0,0 +1,680 @@
+local _ENV = mkmodule("plugins.buildingplan.unlink_mechanisms")
+
+local gui = require("gui")
+local overlay = require("plugins.overlay")
+local utils = require("utils")
+local widgets = require("gui.widgets")
+
+saved_mode = saved_mode or 0
+
+local function mech_iter(b) --iterate mechanisms backwards
+    local t = b.contained_items
+    local i_cur = #t - 1
+    return function()
+        for i = i_cur, 1, -1 do --index 0 is always building component
+            local item = t[i].item
+            if item._type == df.item_trappartsst and
+                item.flags.in_building and
+                not item.flags.in_job then
+                    i_cur = i - 1 --resume here
+                    return item
+            end
+        end
+    end
+end
+
+local function get_trigger_index(m) --return gref index or nil
+    for k, gref in ipairs(m.general_refs) do
+        if gref._type == df.general_ref_building_triggerst or
+            gref._type == df.general_ref_building_triggertargetst then
+                return k
+        end
+    end
+end
+
+local function get_mech_target(m) --mechanism target building if exists
+    local i = get_trigger_index(m)
+    return i and df.building.find(m.general_refs[i].building_id) or nil
+end
+
+local function has_link_tab(b) --linked building tab exists
+    if not b then
+        return
+    elseif b._type == df.building_trapst and
+        (b.trap_type == df.trap_type.Lever or
+        b.trap_type == df.trap_type.PressurePlate) then
+            return true --always active
+    end
+
+    for item in mech_iter(b) do
+        if get_trigger_index(item) then
+            return true
+        end
+    end
+end
+
+local function unlink_building(b, m) --clean up extra stuff
+    if b._type == df.building_trapst then
```

Comment:
we normally write this `df.building_trapst:is_instance(b)`, but this is probably ok (as long as `b` cannot be `nil`)